pyccel/parser/semantic.py

```diff
@@ -1140,57 +1163,120 @@ def _handle_function(self, expr, func, args, is_method = False):
-
-            if None in new_expr.args:
-                errors.report("Too few arguments passed in function call",
-                        symbol = expr,
-                        severity='error')
-            elif isinstance(func, FunctionDef):
-                self._check_argument_compatibility(args, func_args,
-                            expr, func.is_elemental)
             return new_expr
 
+    def _sort_function_call_args(self, func_args, args):
+        """
+        Sort and add the missing call arguments to match the arguments in the function definition.
+
+        Sort and add the missing call arguments to match the arguments in the function definition.
+
+        Parameters
+        ----------
+        func_args : list
+          The function def arguments.
+        args : list
+          The call arguments.
+
+        Returns
+        -------
+        input_args:
+          The sorted and complete call arguments.
+        """
+        input_args = [a for a in args if a.keyword is None]
+        nargs = len(input_args)
+        for ka in func_args[nargs:]:
+            key = ka.name
+            relevant_args = [a for a in args[nargs:] if a.keyword == key]
+            n_relevant_args = len(relevant_args)
+            assert n_relevant_args <= 1
+            if n_relevant_args == 0 and ka.has_default:
+                input_args.append(ka.default_call_arg)
+            elif n_relevant_args == 1:
+                input_args.append(relevant_args[0])
+
+        return input_args
+
+    def _annotate_the_called_function_def(self, func, function_call=None):
+        """
+        Annotate the called FunctionDef.
+
+        Annotate the called FunctionDef.
+
+        Parameters
+        ----------
+        func : FunctionDef|Interface
+           The function that needs to be annotated.
+
+        function_call : list,optional
+           The list of the call arguments.
+
+        Returns
+        -------
+        func: FunctionDef|Interface
+            The new annotated function.
+        """
+        old_scope            = self._scope
+        old_current_function = self._current_function
+        names = []
+        sc = func.scope if isinstance(func, FunctionDef) else func.syntactic_node.scope
+        while sc.parent_scope is not None:
+            sc = sc.parent_scope
+            if not sc.name is None:
+                names.append(sc.name)
+        names.reverse()
+
+        if len(names) == 0:
+            self._current_function = None
+        else:
+            self._current_function = DottedName(*names) if len(names)>1 else names[0]
+
+        while len(names) > 0:
+            sc = sc.sons_scopes[names[0]]
+            names = names[1:]
+        self._scope = sc
+        self._visit_FunctionDef(func, annotate=True, function_call=function_call)
+        user_nodes = func.get_all_user_nodes()
+        func       = self.scope.find(func.name, 'functions')
+        func.get_all_user_nodes().extend(user_nodes)
```

Comment:
I don't remember this line anymore, if might not be needed

Reply:
The problem appears when we import an inlined function, the imported function is cloned, so it does not contain the necessary information that is needed, which is why I add them in this manner.
but yes probably there is a better method in handling this.
we need the nodes to be able to execute this [line](https://github.com/pyccel/pyccel/blob/devel-inline-f/pyccel/codegen/printing/fcode.py#L410)

Reply:
When a `FunctionDef` is created it does not have a user node. The users are added as they are created. So the direct Module user is added here:
https://github.com/pyccel/pyccel/blob/fd1a13e135ff63a6ca878ec50511cc19b1961fd6/pyccel/parser/semantic.py#L2248-L2256
I think you finish visiting functions before building this Module so this would still be ok. The remaining case is then functions in functions. This is handled here:
https://github.com/pyccel/pyccel/blob/fd1a13e135ff63a6ca878ec50511cc19b1961fd6/pyccel/parser/semantic.py#L4155-L4172

I think the possible user nodes of FunctionDef are:
- Module (I think this should work out of the box)
- FunctionDef
- FunctionCall

Do you think this can be achieved with `substitute`? I am also concerned that the user nodes aren't being updated so there could be for example FunctionCalls which point to the wrong FunctionDef. In places where functions are renamed (e.g. for `__init__` and soon for functions printed in C as per #1150) we rely on the fact that in Pyccel we only have 1 object representing each concept in the code.

Reply:
I okay, Im not too familiar with this, maybe to solution is not to clone the imports to that if we change them in one file they will change in the other file

Reply:
I think it will still have the same problem because the imported file is a syntacitc anyways, so not creating a clone will not solve the issue

Reply:
In the case of an InlinedFunction, it will not be annotated, it is a special case, I can add the check to see if the function was imported and inlined, and add the nodes to it

Reply:
At the moment that is what we do, we add the syntactic node to the tree to know that it exists, and then annotate it when the function is called

Reply:
Issue #1728 is going to do the same, we need a way to know that the function exists when we are in the function call before the annotation, and then we annotate it afterwards

Reply:
> At the moment that is what we do, we add the syntactic node to the tree to know that it exists, and then annotate it when the function is called

We add it to the scope but not to the tree of user/attribute nodes.

> Issue https://github.com/pyccel/pyccel/pull/1728 is going to do the same

There is no issue #1728. That link points to this PR. Not sure what you're talking about here

Reply:
ah sorry I mean the issue #1768

Reply:
It should be easy to fix, I will just insert the annotated function in the imports, and this should fix the problem

Reply:
Usually the way to know that the function exists is to add it to the scope, not to add it to the Module though.

> I will just insert the annotated function in the imports, and this should fix the problem

The imports of the Module? This seems like a strange solution. I'm not sure I understand why you want to do this. Won't it break the printing?

Reply:
mmmm, so when we create a Module, it adds itself to all its children, so don't I just need to take all the nodes used in `old_func` which is not annotated, and add them to the `new_func` which is annotated that has an empty user_nodes?

Reply:
let's just say that I annotate the function after the creation of a module here
https://github.com/pyccel/pyccel/blob/fd1a13e135ff63a6ca878ec50511cc19b1961fd6/pyccel/parser/semantic.py#L2248-L2256
what do we need to do to have the proper user_nodes?

Reply:
The problem is not with `user_nodes` (this is _a_ correct way to get the correct user_nodes), rather the problem is with the attribute nodes of the Module. `func` knows that it belongs to `Module`, but `Module` doesn't know that `func` is inside it. Because the two (user_nodes/attribute_nodes) must be set at the same time the normal method for updating these two properties is to add a function to `Module` (e.g. `add_function`) which takes care of setting both properties

Reply:
[a175155](https://github.com/pyccel/pyccel/pull/1728/commits/a175155df12e0cfe33a00b58e8e538d27e6658a9)
this seems to fix the issue, but Module still doesn't know, that the function exists, which is fine because it never gets printed anyway

Reply:
I agree that this feels safer inside an `is_imported:` block.